v1/plugin/stream_collector_proxy.go

```diff
@@ -2,86 +2,130 @@ package plugin
 
 import (
 	"fmt"
+	"time"
 
 	"golang.org/x/net/context"
 
 	"github.com/intelsdi-x/snap-plugin-lib-go/v1/plugin/rpc"
 )
 
+const defaultMaxCollectDuration = 10 * time.Second
+
 type StreamProxy struct {
 	pluginProxy
-
 	plugin StreamCollector
+
+	// maxMetricsBuffer is the maximum number of metrics the plugin should buffer
+	// before sending metrics, defaults to zero what means send metrics immediately
+	maxMetricsBuffer int64
+
+	// maxCollectionDuration sets the maximum duration between collections
+	// before metrics should be sent (defaults to 10s what means that after
+	// 10 seconds, the plugin should send data whatever it has instead of waiting longer)
+	maxCollectDuration time.Duration
 }
 
-func (c *StreamProxy) StreamMetrics(stream rpc.StreamCollector_StreamMetricsServer) error {
+func (p *StreamProxy) StreamMetrics(stream rpc.StreamCollector_StreamMetricsServer) error {
+	// set default max collect duration
+	p.setMaxCollectDuration(time.Duration(defaultMaxCollectDuration))
+
 	// Error channel where we will forward plugin errors to snap where it
 	// can report/handle them.
 	errChan := make(chan string)
 	// Metrics into the plugin from snap.
 	inChan := make(chan []Metric)
 	// Metrics out of the plugin into snap.
 	outChan := make(chan []Metric)
-	err := c.plugin.StreamMetrics(inChan, outChan, errChan)
+
+	err := p.plugin.StreamMetrics(inChan, outChan, errChan)
 	if err != nil {
 		return err
 	}
-	go metricSend(c.plugin, outChan, stream)
-	go errorSend(c.plugin, errChan, stream)
-	streamRecv(c.plugin, inChan, stream)
+
+	go p.metricSend(outChan, errChan, stream)
+	p.streamRecv(inChan, stream)
+
 	return nil
 }
 
-func errorSend(
-	plugin StreamCollector,
-	ch chan string,
-	s rpc.StreamCollector_StreamMetricsServer) {
-	for r := range ch {
-		reply := &rpc.CollectReply{
-			Error: &rpc.ErrReply{Error: r},
-		}
-		if err := s.Send(reply); err != nil {
-			fmt.Println(err.Error())
-		}
+func sendReply(metrics []*rpc.Metric, stream rpc.StreamCollector_StreamMetricsServer) {
+	if len(metrics) == 0 {
+		fmt.Println("No metrics available")
+		return
+	}
+
+	reply := &rpc.CollectReply{
+		Metrics_Reply: &rpc.MetricsReply{Metrics: metrics},
+	}
 
+	if err := stream.Send(reply); err != nil {
```

Comment:
I can add checking if the stream is not a nil in `StreamMetrics()` which is the beginning of the whole flow.